resolwe/flow/models.py

```diff
@@ -479,6 +479,35 @@ def add_dependency(value):
                 for data in value:
                     add_dependency(data)
 
+    def validate_json_fields(self):
+        try:
+            if self.pk is None:  # on create
+                validate_schema(self.input, self.process.input_schema)  # pylint: disable=no-member
+
+            if self.descriptor_schema:
+                validate_schema(self.descriptor, self.descriptor_schema.schema)  # pylint: disable=no-member
+            elif self.descriptor and self.descriptor != {}:
+                raise ValueError("`descriptor_schema` must be defined if `descriptor` is given")
+
+            path_prefix = os.path.join(settings.FLOW_EXECUTOR['DATA_DIR'], str(self.pk))
+            output_schema = self.process.output_schema  # pylint: disable=no-member
+            if self.status == Data.STATUS_DONE:
+                validate_schema(self.output, output_schema, path_prefix=path_prefix)
+            else:
+                validate_schema(self.output, output_schema, path_prefix=path_prefix,
+                                test_required=False)
+        except ValidationError as ex:
+            # if object doesn't have ``pk`` assigned and doesn't validate, it
+            # won't be stored in database, so we don't have to do anything
+            if self.pk:
+                # current object is already corupted, so we have to get fresh
+                # copy (last that has passed validation)
+                data = Data.objects.get(pk=self.pk)
```

Comment:
> as otherwise it should be handled on (too) many places.

Where must this be handled besides the process executor?

> In your example, ValidationError would be (re)raised, and the nice thing about this is that Data object would stay consistent even when updating it through shell ( = status and error would be updated automatically as you probably wouldn't handle it manually).

Yes, this is even stranger. An exception gets raised during `save`, but the object is still mutated? This does not interact well with transaction handling. And I don't see this as consistent - there is no error, because the object was not saved (the save was aborted). But the object now shows an error just because some later operation failed on it (and was aborted)?

So in the general case this looks like some totally unexpected behavior when manipulating a database model. It does make sense when you look at it from the process executor's perspective: the process tried to send invalid data to the executor, so the process itself has been aborted and the Data's status reflects that error condition. But, to me this seems like something that only makes sense when implemented inside the executor and not in any other situation.

But in the general case, the erroneous data may come from anywhere and it is not always the correct way to just mutate the Data object so it now displays an error, related to the process. Perhaps the `save` caller wants to handle the validation error differently?